### PR TITLE
Add minimal WebSocket server implementation for evhttp

### PR DESCRIPTION
**Pull request by widgetii**
This adds few functions to use evhttp-based webserver to handle incoming WebSockets connections. We've tried to use both libevent and libwebsockets in our application, but found that we need to have different ports at the same time to handle standard HTTP and WebSockets traffic. This change can help to stick only with libevent library.

Implementation was inspired by modified Libevent source code in [ipush](https://github.com/sqfasd/ipush/tree/master/deps/libevent-2.0.21-stable) project.

Also, WebSocket-based chat server was added as a sample.

### CODE REVIEW

**Comment by azat** [comment]
Thank you! This is a great feature!

Do you have some performance numbers for websocket server?
Also can you please add some unit tests for this great feature?

No we have stable CI. So could you please fix building `ws-chat-server.c` sample for windows?

```
2022-08-10T09:48:37.9578843Z D:\a\libevent\libevent\sample\ws-chat-server.c(13,10): fatal error C1083: Cannot open include file: 'unistd.h': No such file or directory [D:\a\libevent\libevent\build\ws-chat-server.vcxproj]
```

**Comment by widgetii** [comment]
I've tried to fix all issues I found, but still stuck with proper unit tests. Could you give me a direction to use them right (internal libevent's kitchen is new for me)?

I added two cases: when bad handshake is provided and a regular WS data exchange. After I read out all headers in `http_ws_readcb_hdr` it seems that both sides (server and emulated client) are in some kind of deadlock state

File: test/regress_http.c [hunk 1]
```diff
@@ -1304,6 +1307,206 @@ http_allowed_methods_test(void *arg)
...
+	line = evbuffer_readln(input, &nread, EVBUFFER_EOL_CRLF);
+	if (!strncmp(line, "HTTP/1.1 401 ", strlen("HTTP/1.1 401 "))) {
+		test_ok++;
+	}
+	if (line)
+		free(line);
+}
+
+static void
+http_ws_test(void *arg)
```
Comment (azat) [inline test/regress_http.c:1454]: Now, you should register the test here - https://github.com/libevent/libevent/blob/bb57cea387ba4ebac077c71753e391225adfc81f/test/regress_http.c#L5899

Also consider moving tests into separate file, `test/regress_ws.c` ?

**Comment by azat** [comment]
>I've tried to fix all issues I found, but still stuck with proper unit tests. Could you give me a direction to use them right (internal libevent's kitchen is new for me)?

Maybe this will help? https://github.com/libevent/libevent/pull/1322/files#r947026444

**Comment by widgetii** [comment]
I just skipped including new testes for making all CI process pass (after adding includes for Windows and FreeBSD systems) and now going to revert it. But in the current state WS tests will not pass.

**Comment by widgetii** [comment]
Basically, I'm done and PR is ready for review

File: ws.c [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+}
+
+static char *
+ws_gen_accept_key(const char *ws_key, char out[32])
+{
+	char buf[1024];
+	char digest[20];
+
+	snprintf(
+		buf, sizeof(buf), "%s258EAFA5-E914-47DA-95CA-C5AB0DC85B11", ws_key);
```
Comment (azat) [inline ws.c:138]: Let's add a constant for this (i.e. `WS_UUID`)

File: ws.c [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+				text[msg_len] = 0;
+				evws->cb(evws, text, msg_len, evws->cb_arg);
+				free(text);
+			}
+		} else if (type == INCOMPLETE_TEXT_FRAME || type == INCOMPLETE_FRAME) {
+			/* incomplete frame received, wait for next chunk */
+			return;
+		} else if (type == INCOMPLETE_BINARY_FRAME || type == BINARY_FRAME ||
+				   type == CLOSING_FRAME || type == ERROR_FRAME) {
+			evws_force_disconnect_(evws);
```
Comment (azat) [inline ws.c:285]: So this implementation does not support binary frames. But it looks like pretty easy to implement?

File: test/regress_ws.c [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+	struct evbuffer *output = bufferevent_get_output(bev);
+	size_t nread = 0, n = 0;
+	char *line;
+
+	while ((line = evbuffer_readln(input, &nread, EVBUFFER_EOL_CRLF))) {
+		if (n == 0 &&
+			!strncmp(line, "HTTP/1.1 101 ", strlen("HTTP/1.1 101 "))) {
+			test_ok++;
+		} else if (!strcmp(line,
+					   "Sec-WebSocket-Accept: HSmrc0sMlYUkAGmm5OPpG2HaGWk=")) {
```
Comment (azat) [inline test/regress_ws.c:276]: It is fine to use constant here, but using decode/encode here will be better.

File: test/regress_ws.c [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+	if (!fin) {
+		*options = 1;
+	}
+
+	evbuffer_drain(buf, header_len + payload_len);
+	return out_buf;
+}
+
+static void
+send_ws_msg(struct evbuffer *buf, const char *msg)
```
Comment (azat) [inline test/regress_ws.c:197]: Maybe add some internal helper and use it here instead of doing the same encoding again?

File: test/regress_ws.c [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+{
+	/** For ssl */
+	if (what & BEV_EVENT_CONNECTED)
+		return;
+	test_ok++;
+	event_base_loopexit(arg, NULL);
+}
+
+static char *
+receive_ws_msg(struct evbuffer *buf, size_t *out_len, unsigned *options)
```
Comment (azat) [inline test/regress_ws.c:133]: And same here, maybe add an internal helper?

File: sample/ws-chat.html [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        if (!msg.value) {
+            return false;
+        }
+        conn.send(msg.value);
+        msg.value = "";
+        return false;
+    };
+
+    if (window["WebSocket"]) {
+        conn = new WebSocket("ws://" + document.location.host + "/ws");
```
Comment (azat) [inline sample/ws-chat.html:32]: Add a `8080` port here?

File: sample/ws-chat-server.c [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+nice_addr(const char *addr)
+{
+	if (strncmp(addr, "::ffff:", 7) == 0)
+		addr += 7;
+
+	return addr;
+}
+
+static void
+addr2str(struct sockaddr *sa, char *addr, size_t len)
```
Comment (azat) [inline sample/ws-chat-server.c:114]: Tiny bit - let's add a port into messages too

File: sample/ws-chat-server.c [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+		/* Make sure the length still matches, now that we
+		 * opened the file :/ */
+		perror("fstat");
+		goto err;
+	}
+
+
+	evb = evbuffer_new();
+	evbuffer_add_file(evb, fd, 0, st.st_size);
+	close(fd);
```
Comment (azat) [inline sample/ws-chat-server.c:180]: `close` is redundant, it will be done by libevent, see `EVBUF_FS_CLOSE_ON_FREE`

**Review by azat — COMMENTED** [review]
Thank you! The code looks good!

Can you please address the comments and I will merge it.
Most of them are minor, but there are few important (about extra malloc, that can be avoided), and also it will be great to put more comments into the code, plus some summary comment for the implementation (it's limitations)

File: ws.c [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+
+		type = get_ws_frame(data, in_len, &payload, &msg_len);
+
+		if (type == TEXT_FRAME) {
+			if (evws->cb) {
+				char *text = mm_malloc(msg_len + 1);
+
+				memcpy(text, payload, msg_len);
+				text[msg_len] = 0;
+				evws->cb(evws, text, msg_len, evws->cb_arg);
```
[the remark this answers is not in the document]
Comment (widgetii) [reply ws.c:277]: Could you give me an example how to properly use it in this specific case? Previously it was Linux-only implementation and I used a lot of `alloca()` but it seems is not portable

File: ws.c [hunk 11]
```diff
@@ -0,0 +1,409 @@
...
+	int msg_len, in_len;
+	struct evbuffer *input = bufferevent_get_input(evws->bufev);
+
+	while ((in_len = evbuffer_get_length(input))) {
+		unsigned char *data = evbuffer_pullup(input, in_len);
+
+		type = get_ws_frame(data, in_len, &payload, &msg_len);
+
+		if (type == TEXT_FRAME) {
+			if (evws->cb) {
```
Comment (azat) [inline ws.c:272]: BTW is it possible to not have cb? And if so, does it make sense?

File: include/event2/ws.h [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+ * @param req a request object
+ * @param cb the callback function that gets invoked on receiving message
+ * @param arg an additional context argument for the callback
+ * @return a pointer to a newly initialized WebSocket connection or NULL
+ *   on error
+ * @see evws_close()
+ * */
+EVENT2_EXPORT_SYMBOL
+struct evws_connection *evws_new_session(struct evhttp_request *req,
+	void (*cb)(struct evws_connection *, char *, size_t, void *), void* arg);
```
Comment (azat) [inline include/event2/ws.h:22]: Let's add a `typedef` for `cb`, and use it here and in `wc.c`

File: ws.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (azat) [reply ws.c:277]: >Could you give me an example how to properly use it in this specific case?

Actually I don't think that `evbuffer` is good idea here. How about passing `const char *` (to add a sign to user that the data is const/immutable), the reference to existing array, instead of allocating new?

TL;DR;

I thought about some variants, and I cannot find better way than simply pass `const char *`.

Creating temporary `evbuffer` to pass to the callback:
- too costly, 144 (evbuffer) + 48 (evbuffer_chain) is big overhead
- and even if you will ignore overhead, it still does not have any benefits over `const char *`, since if it will hold the reference, then anyway if user want to, he could change it. By the same reason per-connection `evbuffer` will not help.

But just to show you idea, here is a hacky patch -
[ws-evbuffer.patch.txt](https://github.com/libevent/libevent/files/9455518/ws-evbuffer.patch.txt)

> Previously it was Linux-only implementation and I used a lot of alloca() but it seems is not portable

`alloca` is indeed not good to use here.

[the remark this answers is not in the document]
Comment (widgetii) [reply ws.c:277]: Now I got it, my previous idea was to provide a simple API for library user to get WS text message with null terminator and then user just can use it as a const char* in its code
Obviously, there is no way to do this without additional allocations, but for me using `struct evbuffer *` in message callback looks overcomplicated (because of necessity of using `evbuffer_pullup` to extract actual data pointer where we can use just raw pointer and message data length - there is no way of delivering incomplete WS message now)

File: test/regress_ws.c (hunk 4 above)
Comment (widgetii) [reply test/regress_ws.c:276]: I used these "well-known" WS protocol handshake from https://en.wikipedia.org/wiki/WebSocket#Protocol_handshake and thought it would be better to have them in strings for tests to "prove" protocol correctness

File: test/regress_ws.c (hunk 6 above)
Comment (widgetii) [reply test/regress_ws.c:133]: Didn't get how the sending and receiving routines can share both the same code via helpers.

File: ws.c (hunk 11 above)
Comment (widgetii) [reply ws.c:272]: Didn't get this part, what do you mean by that? Libevent somehow needs to deliver parsed messages to the user, and I guess the callback is the most common way to do this. The same idea is used in WebSocket browser API https://developer.mozilla.org/en-US/docs/Web/API/WebSocket

File: test/regress_ws.c (hunk 4 above)
Comment (azat) [reply test/regress_ws.c:276]: Good explanation, agree.

File: ws.c (hunk 11 above)
Comment (azat) [reply ws.c:272]: Seems that I badly explained it, I meant the condition itself `evws->cb`, is it possible for it to be `false`? And if so, does this makes sense?

IOW I'm suggesting removing it, but not sure that it is safe now.

File: test/regress_ws.c (hunk 6 above)
Comment (azat) [reply test/regress_ws.c:133]: I meant not between sending and receiving, to add a common internal helper and use it in `ws.c` and `regress_ws.c`

Comment (widgetii) [reply test/regress_ws.c:133]: But in this case it will be bad in case we have common error in both test and implementation

File: http.c [hunk 13]
```diff
@@ -3191,6 +3192,31 @@ evhttp_send_reply_chunk_with_cb(struct evhttp_request *req, struct evbuffer *dat
...
 
+struct bufferevent *
+evhttp_start_ws_(struct evhttp_request *req)
+{
+	struct evhttp_connection *evcon = req->evcon;
+	struct bufferevent *bufev;
+
+	evhttp_response_code_(req, HTTP_SWITCH_PROTOCOLS, "Switching Protocols");
+
+	if (req->evcon == NULL)
```
[the remark this answers is not in the document]
Comment (widgetii) [reply http.c:3203]: I've implemented `evhttp_start_ws_` using ideas from `evhttp_send_reply_start`, the same `evhttp_response_code_()` call and with the same `req->evcon == NULL` condition. If we don't check `req->evcon` then next call to `evhttp_make_header()` may segfault

File: ws.c (hunk 11 above)
Comment (widgetii) [reply ws.c:272]: Right, it's redundant because is set in `evws_new_session` and there is no way to avoid this set, removing

File: ws.c [hunk 14]
```diff
@@ -0,0 +1,409 @@
...
+	}
+
+	opcode = in_buffer[0] & 0x0F;
+	fin = (in_buffer[0] >> 7) & 0x01;
+	masked = (in_buffer[1] >> 7) & 0x01;
+
+	payload_len = 0;
+	pos = 2;
+	length_field = in_buffer[1] & (~0x80);
+	mask = 0;
```
[the remark this answers is not in the document]
Comment (widgetii) [reply ws.c:209]: I also just checked RFC:
```
Mask:  1 bit

      Defines whether the "Payload data" is masked.  If set to 1, a
      masking key is present in masking-key, and this is used to unmask
      the "Payload data" as per [Section 5.3](https://www.rfc-editor.org/rfc/rfc6455#section-5.3).  All frames sent from
      client to server have this bit set to 1.
```

It seems that masked data are obligatory only while sending from client to server.

File: test/regress_ws.c (hunk 6 above)
Comment (azat) [reply test/regress_ws.c:133]: I'm fine with leave it as-is, just wanted to reduce amount of code.

File: http.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (azat) [reply http.c:3203]: Ok, let's leave as-is.

File: ws.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (widgetii) [reply ws.c]: Just tested [ws-evbuffer.patch.txt](https://github.com/libevent/libevent/files/9455518/ws-evbuffer.patch.txt)`, this not will work because actual input buffer can contain multiple frames at the same time (that why we're using `while` loop)

[the remark this answers is not in the document]
Comment (widgetii) [reply ws.c]: Please check latest implementation, I used your evbuffer ideas (first it looked complicated, but I found code had no proper support for handling incomplete frames), implemented as well with binary frames support

File: ws.c [hunk 15]
```diff
@@ -0,0 +1,433 @@
...
+			break;
+		case PING_FRAME:
+		case PONG_FRAME:
+			/* ping or pong frame */
+			break;
+		default:
+			event_warn("%s: unexpected frame type %d\n", __func__, type);
+			evws_force_disconnect_(evws);
+		}
+		evbuffer_drain(input, msg_len);
```
Comment (azat) [inline ws.c:319]: Should this be done in case of incomplete frames?

File: include/event2/ws.h [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+#define WS_BINARY_FRAME 0x2
+
+typedef void (*ws_on_msg_cb)(
+	struct evws_connection *, int type, struct evbuffer *, size_t, void *);
+typedef void (*ws_on_close_cb)(struct evws_connection *, void *);
+
+/** Opens new WebSocket session from HTTP request.
+ *
+ * @param req a request object
+ * @param cb the callback function that gets invoked on receiving message
```
Comment (azat) [inline include/event2/ws.h:21]: Let's describe the behavior of the buffer, i.e. that user should not modify it, and libevent will remove processed bytes by itself.

File: ws.c [hunk 17]
```diff
@@ -0,0 +1,433 @@
...
+ws_evhttp_read_cb(struct bufferevent *bufev, void *arg)
+{
+	struct evws_connection *evws = arg;
+	unsigned char *payload;
+	enum WebSocketFrameType type;
+	int msg_len, in_len, header_sz;
+	struct evbuffer *input = bufferevent_get_input(evws->bufev);
+
+	while ((in_len = evbuffer_get_length(input))) {
+		unsigned char *data = evbuffer_pullup(input, in_len);
```
Comment (azat) [inline ws.c:273]: May fail

**Review by azat — APPROVED** [review]
Generally LGTM. Just some minor comments.

File: ws.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (azat) [reply ws.c]: That what we need, though a little bit unsafe, but I guess it is fine!

File: ws.c (hunk 15 above)
Comment (widgetii) [reply ws.c:319]: Incomplete frames will not pass through here because of `continue` operator which do next iteration for the loop

File: ws.c (hunk 17 above)
Comment (widgetii) [reply ws.c:273]: Do you mean fail of `evbuffer_pullup` (I know that it may return NULL, but in the previous step we determine `in_len` and pullup exactly `in_len` size)?

Comment (azat) [reply ws.c:273]: Yes, failure of `evbuffer_pullup()`. It may be possible because of pinned chains for example, this should not happens here, but it should be definitely handled anyway.

File: include/event2/ws.h [hunk 18]
```diff
@@ -16,14 +16,18 @@ typedef void (*ws_on_msg_cb)(
...
- * @param cb the callback function that gets invoked on receiving message
- * @param arg an additional context argument for the callback
- * @return a pointer to a newly initialized WebSocket connection or NULL
- *   on error
- * @see evws_close()
- * */
+  @param req a request object
+  @param cb the callback function that gets invoked on receiving message
+  with evbuffer with len bytes of the incoming data available in
+  it. User may use `evbuffer_pullup` to get a pointer to the data, but shouldn't
```
Comment (azat) [inline include/event2/ws.h:22]: Actually I thought about this a little bit more and now it became too unsafe, since user may still call some remove/drain/..., what do you think about passing `const char *` to callback instead of buffer?

Comment (widgetii) [reply include/event2/ws.h:22]: As for me, first approach (without passed evbuffer to the user) was more clear and user-friendly (if it even could be applied to users of libevent library)

The main issue is that user except `const char *` with terminated \0 for text messages, but I have no idea how to implement this without additional memory allocation and copying

Comment (azat) [reply include/event2/ws.h:22]: Okay, convinced, let's get back to an approach with a separate allocations.
Sorry for the noise.

> The main issue is that user except const char * with terminated \0 for text messages

This is not a an issue, it just should be documented, `const char *` is not always NUL terminated.

Comment (widgetii) [reply include/event2/ws.h]: Fixed

**Comment by azat** [comment]
@widgetii many thanks for you contribution!

**Comment by kaend** [comment]
Tor fails to build with this .
Conflict with `SHA1' function from openssl
```
  CCLD     src/app/tor.exe
D:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: D:/msys64/home/user/zx/lib/libcrypto.a(libcrypto-lib-sha1_one.obj):sha1_one.c:(.text+0x80): multiple definition of `SHA1'; D:/msys64/home/user/zx/lib/libevent.a(sha1.o):sha1.c:(.text+0x16e0): first defined here
collect2.exe: error: ld returned 1 exit status
```

**Comment by widgetii** [comment]
There are two options:
1) Rename our `SHA1` into something with different name
2) Don't use shipped sha1 routines when build with OpenSSL or MbedTLS, but rather use their own implementations

@azat What do you think is better?

**Comment by azat** [comment]
Let's just rename bundled.

**Comment by liudongmiao** [comment]
There is no frame type for `evws_send`, I'd suggest at least two type, text and binary.

**Comment by liudongmiao** [comment]
I'd suggest the `reason` of `evws_close` use enum. And actually, RFC 6445 defines optional message for close frame.

**Comment by liudongmiao** [comment]
The name of `WebSocketFrameType` should be avoided. The main code style for enum in libevent seems lower case, and some in rpc is upper case.

And, for the function `make_ws_frame`, it seems no mask, so, it's only for server response, then it should be renamed to `make_server_ws_frame` or similar.
If it's only for server response, then header[16] is useless.
Actually, for client, it's max to 14, for server, as there is no mask, it's max to 10.
I'd suggest use union or similar, then use `ntohs` for 16-bits length and `ntohll` for 64-bits length.
```c
header[pos++] = (unsigned char)frame_type | 0x80; /* fin */
```
It should be `0x80 | ((unsigned char)frame_type & 0xf)`, avoid frame_type for override high bit `fin:1 rsv:3`.
// In our case, we never use 64 bits length, and in most case, a frame of 16 bits (64K) should be enough.

And, there is no response for ping / pong. Response for pong is useless, however, Response to ping should be handled.

BTW, I use `bufferevent_filter_new` to handle websocket frame. I love `bufferevent_filter_new`, specially for `BEV_NEED_MORE`.

**Comment by widgetii** [comment]
This PR was merged more than 1,5 years ago and some changes were done on top of it (like differentiations between text and binary data)

If you have any improvements to the code, fell free to open new PR
